### PR TITLE
mips: enable SLJIT_HAVE_REV for SLJIT_MIPS_REV >= 2

### PR DESCRIPTION
**Pull request by carenas**
Adds 3 instructions from "revision 2" that allow for a new implementation of `SLJIT_REV` (and friends) that should be good enough to enable the corresponding `sljit_has_cpu_feature()` flag.

[documentation](https://github.com/zherczeg/sljit/blob/3dcdb25754363db2cd34dbd1cd9a6a6b880cadae/doc/overview.txt#L41-L44) for what is an invalid operation has been also updated.

### CODE REVIEW

**Comment by zherczeg** [comment]
Could you explain the problem with the test?

**Comment by carenas** [comment]
> Could you explain the problem with the test?

`ibuf[4]` related code is missing, but more importantly the original case 12 sets junk in the upper half of the register instead of the full sign extension that would be expected, if it is meant to be truncated and used later as a 32bit register.

I understand that being a test is trying to trigger edge cases, but it goes against the "principle" that defines all other sljit operations as valid, and it requires additional code to guard against.

**Comment by zherczeg** [comment]
https://github.com/zherczeg/sljit/blob/master/sljit_src/sljitLir.h#L1106

It looks like rev32 handles 32 bit input, the rest should handle word sized input. So no sign extension is needed.

**Comment by carenas** [comment]
not sure I understand your comment, but the problem isn't in sljit code but the fact that the instructions in MIPS have a restriction that says:

```
If GPR rt does not contain a sign-extended 32-bit value (bits 63..31 equal), then the result of the operation is
UNPREDICTABLE.
```

I added a case for test89 that hopefully helps clarify the issue and that does NOT fail on either QEMU or the Octeon CPU I had been testing with, and my assumption is that hopefully "UNPREDICTABLE" means that since the sign is ambiguous when the CPU completes the operation and does the sign extension of the result to put it back in the destination register it might get it wrong. Luckily we already have to "fix" (MIPS always sign extends) the sign extension anyway for `SLJIT_REV_{U,S}16` so hopefully that would help there.

**Comment by carenas** [comment]
in a related theme, is `SLJIT_REV_S16` and other signed variants really needed?

**Comment by zherczeg** [comment]
There is a reason we have `SLJIT_REV_U16` and `SLJIT_REV32_U16`. The former must not expect that the value is sign extended or anything. The latter can expect that the upper 32 bit is the sign extension of bit 31 (on MIPS). The signed/unsigned forms should be useful for loading signed values.

**Comment by carenas** [comment]
My point was, why would you use `SLJIT_REV_S16` or `SLJIT_REV_S32`, if it is semantically better to use `SLJIT_REV_U16` or `SLJIT_REV_U32`.

For example, `gcc` does provide builtins for all `bswap` functions, but they are only expected to work with unsigned values, and so the whole "sign extension" issue is irrelevant.

in MIPS we will then set the upper word to 0 initially and any "automatic" sign extension coul be safely ignored.

**Comment by zherczeg** [comment]
I mentioned above the primary use case is loading signed reversed values, which can be optimized depending on the cpu capabilities.

File: test_src/sljitTest.c [hunk 1]
```diff
@@ -11288,37 +11301,57 @@ static void test90(void)
 
 	sljit_emit_enter(compiler, 0, SLJIT_ARGS2(VOID, W, W), 5, 5, 0, 0, 2 * sizeof(sljit_sw));
 
-	sljit_emit_op1(compiler, SLJIT_MOV, SLJIT_R0, 0, SLJIT_IMM, WCONST(0xffffa1b2c3d4, 0xa1b2c3d4));
```
Comment (zherczeg) [inline test_src/sljitTest.c:11291]: This is not need to be changed

Comment (carenas) [reply test_src/sljitTest.c:11291]: OK, but then the "description" of the case should change, or even better a different case which specifically handles the "junk" in the upper word is ignored by this instruction should be added instead IMHO

File: test_src/sljitTest.c [hunk 2]
```diff
@@ -11339,11 +11376,6 @@ static void test90(void)
 	/* ibuf[4] */
 	ibuf[4] = -824;
 
-	sljit_emit_op1(compiler, SLJIT_MOV, SLJIT_R4, 0, SLJIT_IMM, WCONST(0xffff8192a3b4, 0x8192a3b4));
-	sljit_emit_op1(compiler, SLJIT_REV_S32, SLJIT_S4, 0, SLJIT_R4, 0);
-	/* buf[6] */
-	sljit_emit_op1(compiler, SLJIT_MOV, SLJIT_MEM1(SLJIT_S0), 6 * sizeof(sljit_sw), SLJIT_S4, 0);
-
```
Comment (carenas) [inline test_src/sljitTest.c:11346]: maybe this test should had been `IS_64BIT` and focus on the "store" of a truncated 32bit signed instead?

see the case added to test89 for details.

File: test_src/sljitTest.c (hunk 1 above)
Comment (zherczeg) [reply test_src/sljitTest.c:11291]: The instruction reverses the lower 4 bytes, and sign/unsign extend the result. The upper 4 byte is ignored, since its value is not specified. The value is usually the result of some computation, and the upper bits depend on that. There is a REV32 variant, which uses the result of a 32 bit computation, so the upper bits might be defined by the cpu.

File: GNUmakefile [hunk 3]
```diff
@@ -16,7 +16,7 @@ EXTRA_LDFLAGS=
 endif
 
 CPPFLAGS = $(EXTRA_CPPFLAGS) -Isljit_src
-CFLAGS += -O2 -Wall -Wextra -Wconversion -Wsign-compare -Werror
+CFLAGS += -g -O0 -Wall -Wextra -Wconversion -Wsign-compare -Werror
```
Comment (zherczeg) [inline GNUmakefile:19]: Keep the original.

File: test_src/sljitTest.c [hunk 4]
```diff
@@ -11262,6 +11275,12 @@ static void test89(void)
 	FAILED(hbuf[5] != (sljit_s16)0x0201, "test89 case 14 failed\n");
 	FAILED(hbuf[6] != (sljit_s16)0x0201, "test89 case 15 failed\n");
 	FAILED(hbuf[7] != -367, "test89 case 16 failed\n");
+#if IS_64BIT
+	FAILED(hbuf[8] != 0x55aa, "test89 case 17 failed\n");
+	FAILED(buf[5] != hbuf[8], "test89 case 17 (zero extension) failed\n");
```
Comment (zherczeg) [inline test_src/sljitTest.c:11280]: this can be 18, 19,...

File: test_src/sljitTest.c [hunk 5]
```diff
@@ -11272,51 +11291,59 @@ static void test90(void)
...
 
 	sljit_emit_enter(compiler, 0, SLJIT_ARGS2(VOID, W, W), 5, 5, 0, 0, 2 * sizeof(sljit_sw));
 
+	/* zero extend any integer */
 	sljit_emit_op1(compiler, SLJIT_MOV, SLJIT_R0, 0, SLJIT_IMM, WCONST(0xffffa1b2c3d4, 0xa1b2c3d4));
 	sljit_emit_op1(compiler, SLJIT_REV_U32, SLJIT_R0, 0, SLJIT_R0, 0);
 	/* buf[0] */
 	sljit_emit_op1(compiler, SLJIT_MOV, SLJIT_MEM1(SLJIT_S0), 0, SLJIT_R0, 0);
 
-	sljit_emit_op1(compiler, SLJIT_MOV, SLJIT_R2, 0, SLJIT_IMM, WCONST(0xffffa1b2c3d4, 0xa1b2c3d4));
```
Comment (zherczeg) [inline test_src/sljitTest.c:11296]: All similar tests should be kept.

**Review by zherczeg — CHANGES_REQUESTED** [review]

File: test_src/sljitTest.c (hunk 4 above)
Comment (carenas) [reply test_src/sljitTest.c:11280]: but using a few extra characters, provide additional information on what part of the behaviour is broken, as well as clarifying that the incorrect value was generated as part of a single "case" code block and therefore aid with troubleshooting.

File: test_src/sljitTest.c (hunk 5 above)
Comment (carenas) [reply test_src/sljitTest.c:11296]: This test is still doing the same, with the only caveat that is now no longer undefined, because the value stored in the register matches what the operation (SLJIT_REV_S32) expects regarding sign extension of the upper 32bit.

The special case were that is not undefined is now covered by the improved case12 that does so by using a MEM destination and ensuring the relevant value was written and that write didn't overflow.

File: test_src/sljitTest.c (hunk 4 above)
Comment (zherczeg) [reply test_src/sljitTest.c]: I am ok with the characters, just increase the numbers.

File: sljit_src/sljitNativeMIPS_common.c [hunk 6]
```diff
@@ -1671,25 +1694,34 @@ static sljit_s32 emit_rev(struct sljit_compiler *compiler, sljit_s32 op, sljit_s
...
+#endif /* SLJIT_MIPR_REV >= 2 */
 	return SLJIT_SUCCESS;
 }
 
 static sljit_s32 emit_rev16(struct sljit_compiler *compiler, sljit_s32 op, sljit_s32 dst, sljit_sw src)
 {
+#if defined(SLJIT_MIPS_REV) && SLJIT_MIPS_REV >= 2
+#if defined(SLJIT_CONFIG_MIPS_32) && SLJIT_CONFIG_MIPS_32
+	FAIL_IF(push_inst(compiler, WSBH | T(src) | D(dst), DR(dst)));
+#else
```
Comment (zherczeg) [inline sljit_src/sljitNativeMIPS_common.c:1711]: comment `/* !SLJIT_CONFIG_MIPS_32 */` missing

File: test_src/sljitTest.c [hunk 7]
```diff
@@ -11293,30 +11316,37 @@ static void test90(void)
 	/* buf[0] */
 	sljit_emit_op1(compiler, SLJIT_MOV, SLJIT_MEM1(SLJIT_S0), 0, SLJIT_R0, 0);
 
+	/* sign extend negative integer */
```
Comment (zherczeg) [inline test_src/sljitTest.c:11319]: Please use full sentences/
/* Sign extend negative integer. */
Similar cases below.

File: test_src/sljitTest.c [hunk 8]
```diff
@@ -11354,16 +11383,17 @@ static void test90(void)
...
+	FAILED(buf[2] != 0x4d3c2b1a, "test90 case 3 failed\n");
+	FAILED(buf[3] != WCONST(0xc6d7e8f9, 0xc6d7e8f9), "test90 case 4 failed\n");
+	FAILED(buf[4] != WCONST(-0x39281707, 0xc6d7e8f9), "test90 case 5 failed\n");
+	FAILED(buf[5] != WCONST(0xddccbbaa, 0xddccbbaa), "test90 case 6 failed\n");
+	FAILED(buf[6] != WCONST(-0x22334456, 0xddccbbaa), "test90 case 7 failed\n");
+	FAILED(ibuf[0] != 0x04030201, "test90 case 8 failed\n");
+	FAILED(ibuf[1] != 0x04030201, "test90 case 9 failed\n");
+	FAILED(ibuf[2] != (sljit_s32)0xc0d0e0f0, "test90 case 10 failed\n");
+	FAILED(ibuf[3] != (sljit_s32)0xc0d0e0f0, "test90 case 11 failed\n");
+	FAILED(ibuf[4] != (sljit_s32)0xb4a39281, "test90 case 13 failed\n");
```
Comment (zherczeg) [inline test_src/sljitTest.c:11395]: 13 after 11, so 12 is missing

**Review by zherczeg — CHANGES_REQUESTED** [review]
Only some style issues remained

**Review by zherczeg — APPROVED** [review]
LGTM
